src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp

```diff
@@ -313,6 +313,50 @@ void runtime::gpu::CudaKernelBuilder::get_pad_dynamic_op(
     writer.block_end();
 }
 
+void runtime::gpu::CudaKernelBuilder::get_reverse_sequence_op(
+    codegen::CodeWriter& writer,
+    const std::string& name,
+    const std::array<std::string, 3>& data_types,
+    const size_t batch_axis,
+    const size_t sequence_axis,
+    const size_t rank)
+{
+    writer << "extern \"C\" __global__ void cuda_" << name << "(" << data_types[0] << "* in, "
+           << data_types[1] << "* sequence, " << data_types[2] << "* out, "
+           << "uint32_t* output_shape, uint32_t* output_strides, uint32_t n)\n";
+    writer.block_begin();
+    {
+        writer << "uint32_t tid = blockIdx.x * blockDim.x + threadIdx.x;\n";
+        writer << "if (tid < n)\n";
+        writer.block_begin();
+        {
+            writer << "uint32_t input_idx = tid;\n";
+            for (size_t i = 0; i < rank; i++)
+            {
+                writer << "uint32_t output_idx_" << i << " = input_idx / output_strides[" << i
+                       << "];\n";
+                writer << "input_idx %= output_strides[" << i << "];\n";
+            }
+            writer << "uint32_t sequence_length = sequence[output_idx_" << batch_axis << "];\n";
+            writer << "assert(sequence_length <= output_shape[" << sequence_axis << "]);\n";
```

Comment:
This assertion is troublesome, if it's triggered, the kernel will only print a warning and the JIT will continue to execute giving incorrect results.

This would be a good case to have different behavior in production vs debug mode, I think. In debug mode, we could emit a check on the host (by DtH coping the contents of sequence out and comparing to the shape).

I'm not sure what the solution is here, perhaps we can leave it for now, but it does make me uncomfortable.

Reply:
It will throw since we use CUDA_SAFE_CALL. something like this:
```
[ RUN      ] GPU.reverse_sequence_n2c3h4w2
ngraph.cu:24: void cuda_reverse_sequence_int32_t_int32_t_int32_t_bi_2_si_1_r_4(signed int *, signed int *, signed int *, unsigned int *, unsigned int *, unsigned int): block: [42,0,0], thread: [0,0,0] Assertion `sequence_length <= output_shape[1]` failed.
ngraph.cu:24: void cuda_reverse_sequence_int32_t_int32_t_int32_t_bi_2_si_1_r_4(signed int *, signed int *, signed int *, unsigned int *, unsigned int *, unsigned int): block: [18,0,0], thread: [0,0,0] Assertion `sequence_length <= output_shape[1]` failed.
ngraph.cu:24: void cuda_reverse_sequence_int32_t_int32_t_int32_t_bi_2_si_1_r_4(signed int *, signed int *, signed int *, unsigned int *, unsigned int *, unsigned int): block: [11,0,0], thread: [0,0,0] Assertion `sequence_length <= output_shape[1]` failed.
ngraph.cu:24: void cuda_reverse_sequence_int32_t_int32_t_int32_t_bi_2_si_1_r_4(signed int *, signed int *, signed int *, unsigned int *, unsigned int *, unsigned int): block: [43,0,0], thread: [0,0,0] Assertion `sequence_length <= output_shape[1]` failed.
ngraph.cu:24: void cuda_reverse_sequence_int32_t_int32_t_int32_t_bi_2_si_1_r_4(signed int *, signed int *, signed int *, unsigned int *, unsigned int *, unsigned int): block: [35,0,0], thread: [0,0,0] Assertion `sequence_length <= output_shape[1]` failed.
ngraph.cu:24: void cuda_reverse_sequence_int32_t_int32_t_int32_t_bi_2_si_1_r_4(signed int *, signed int *, signed int *, unsigned int *, unsigned int *, unsigned int): block: [27,0,0], thread: [0,0,0] Assertion `sequence_length <= output_shape[1]` failed.
ngraph.cu:24: void cuda_reverse_sequence_int32_t_int32_t_int32_t_bi_2_si_1_r_4(signed int *, signed int *, signed int *, unsigned int *, unsigned int *, unsigned int): block: [26,0,0], thread: [0,0,0] Assertion `sequence_length <= output_shape[1]` failed.
ngraph.cu:24: void cuda_reverse_sequence_int32_t_int32_t_int32_t_bi_2_si_1_r_4(signed int *, signed int *, signed int *, unsigned int *, unsigned int *, unsigned int): block: [19,0,0], thread: [0,0,0] Assertion `sequence_length <= output_shape[1]` failed.
ngraph.cu:24: void cuda_reverse_sequence_int32_t_int32_t_int32_t_bi_2_si_1_r_4(signed int *, signed int *, signed int *, unsigned int *, unsigned int *, unsigned int): block: [10,0,0], thread: [0,0,0] Assertion `sequence_length <= output_shape[1]` failed.
ngraph.cu:24: void cuda_reverse_sequence_int32_t_int32_t_int32_t_bi_2_si_1_r_4(signed int *, signed int *, signed int *, unsigned int *, unsigned int *, unsigned int): block: [34,0,0], thread: [0,0,0] Assertion `sequence_length <= output_shape[1]` failed.
ngraph.cu:24: void cuda_reverse_sequence_int32_t_int32_t_int32_t_bi_2_si_1_r_4(signed int *, signed int *, signed int *, unsigned int *, unsigned int *, unsigned int): block: [3,0,0], thread: [0,0,0] Assertion `sequence_length <= output_shape[1]` failed.
ngraph.cu:24: void cuda_reverse_sequence_int32_t_int32_t_int32_t_bi_2_si_1_r_4(signed int *, signed int *, signed int *, unsigned int *, unsigned int *, unsigned int): block: [2,0,0], thread: [0,0,0] Assertion `sequence_length <= output_shape[1]` failed.
unknown file: Failure
C++ exception with description "
error: cuCtxSynchronize() failed with error
file: /localdisk/tfl/workspace/private-ngraph-cpp/src/ngraph/runtime/gpu/cuda_emitter.cpp
line: 407
msg: CUDA_ERROR_ASSERT" thrown in the test body.
```

Reply:
Ah, true. All right, in that case I think it's fine this way.